firewood/src/merkle/stream.rs

```diff
@@ -88,202 +104,180 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
-                let compare_op = if first_loop {
-                    <u8 as PartialOrd>::ge // >=
-                } else {
-                    <u8 as PartialOrd>::gt
-                };
-
-                let children = get_children_iter(branch)
-                    .filter(move |(_, child_pos)| compare_op(child_pos, &pos));
-
-                let found_next_node =
-                    next_node(merkle, children, visited_path, &mut node, &mut pos)?;
-
-                if !found_next_node {
-                    return Ok(None);
-                }
-            }
-
-            New(NodeType::Branch(branch)) => {
-                if let Some(value) = branch.value.as_ref() {
-                    let value = value.to_vec();
-                    return Ok(Some((node.into_node(), value)));
-                }
-
-                let children = get_children_iter(branch);
+                    // [node_addr] is the next node to visit.
+                    // It's the child at index [pos] of [node_iter].
+                    let Some((node_addr, pos)) = branch_iter.children_iter.next() else {
+                        // We visited all this node's descendants.
+                        // Go back to its parent.
+                        continue;
+                    };
 
-                let found_next_node =
-                    next_node(merkle, children, visited_path, &mut node, &mut pos)?;
+                    let node = merkle
+                        .get_node(node_addr)
+                        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                if !found_next_node {
-                    return Ok(None);
+                    let mut child_key_nibbles = branch_iter.key_nibbles.clone(); // TODO reduce¸cloning
+                    child_key_nibbles.push(pos);
+
+                    branch_iter_stack.push(branch_iter);
+
+                    match node.inner() {
+                        NodeType::Branch(branch) => {
+                            // [children_iter] returns (child_addr, pos)
+                            // for every non-empty child in [node] where
+                            // [pos] is the child's index in [node.children].
+                            let children_iter = get_children_iter(branch);
+
+                            branch_iter_stack.push(BranchIterator {
+                                key_nibbles: child_key_nibbles.clone(), // TODO reduce¸cloning
+                                children_iter: Box::new(children_iter),
+                            });
+
+                            // If there's a value, return it.
+                            if let Some(value) = branch.value.as_ref() {
+                                let value = value.to_vec();
+                                return Poll::Ready(Some(Ok((
+                                    key_from_nibble_iter(child_key_nibbles.into_iter().skip(1)), // skip the sentinel node leading 0
```

Comment:
Can we just construct `child_key_nibbles` without the leading 0 in the first place?

Reply:
We'd have to special case so that we don't push the 0 index that points from the sentinel node to the root but it's certainly possible